### PR TITLE
Update heuristics for AutoML text classification when tune_for_memory set True

### PR DESCRIPTION
**Pull request by amholler**
If AutoML is invoked w/tune_for_memory True on a text model estimated not to fit
in memory even after hyperparameter range reduction, attempt to further reduce
the model size, by replacing Bert with Distilbert, and input sequence length, by
setting word_sequence_length_limit to the minimum of the 99th percentile of
length distribution and 128 (a commonly used max length for BERT on GPUs).

Tested by running AutoML text on the heuristics datasets (agnews dbpedia
ethos_binary goemotions irony sst2 sst3 sst5 yelp_polarity), all of which now
produce a trained model within their time budgets https://docs.google.com/spreadsheets/d/1c1ghzlNBGH8Sh0AgxZw8Z366NZegiblxOuwQB9eHHf4/edit#gid=1666484048

### CODE REVIEW

**Review by justinxzhao — COMMENTED** [review]
Mostly LGTM! Just a couple of comments around testing and readability.

File: ludwig/automl/auto_tune_config.py [hunk 1]
```diff
@@ -153,6 +183,9 @@ def memory_tune_config(config, dataset):
         else:
             param_list.pop(0)  # param not in hyperopt search space
 
+    if (not fits_in_memory) and (model_category == TEXT):
```
[the remark this answers is not in the document]
Comment (amholler) [reply ludwig/automl/auto_tune_config.py]: done

File: ludwig/automl/auto_tune_config.py [hunk 2]
```diff
@@ -98,8 +108,28 @@ def get_new_params(current_param_values, hyperparam_search_space, params_to_modi
     return current_param_values
 
 
+def _reduce_text_model_size(config, training_set_metadata):
```
[the remark this answers is not in the document]
Comment (amholler) [reply ludwig/automl/auto_tune_config.py]: done

File: ludwig/automl/auto_tune_config.py [hunk 3]
```diff
@@ -98,8 +108,28 @@ def get_new_params(current_param_values, hyperparam_search_space, params_to_modi
...
+    logging.info("Text model may overflow memory; reducing model size and input sequence length")
+    min_99ptile_len = AUTOML_SMALLER_TEXT_LENGTH
+    for feature in config["input_features"]:
+        if feature["type"] == TEXT and feature["encoder"] == AUTOML_DEFAULT_TEXT_ENCODER:
+            feature["encoder"] = AUTOML_SMALLER_TEXT_ENCODER
+            feature_99ptile_len = training_set_metadata[feature["name"]]["word_99ptile_max_sequence_length"]
+            if feature_99ptile_len < min_99ptile_len:
+                min_99ptile_len = feature_99ptile_len
```
[the remark this answers is not in the document]
Comment (amholler) [reply ludwig/automl/auto_tune_config.py]: done

File: ludwig/automl/auto_tune_config.py [hunk 4]
```diff
@@ -98,8 +108,28 @@ def get_new_params(current_param_values, hyperparam_search_space, params_to_modi
...
+    seq_len_limit = {"word_sequence_length_limit": round(min_99ptile_len)}
+    if "preprocessing" not in config:
+        config["preprocessing"] = {TEXT: seq_len_limit}
+    elif (
+        (TEXT not in config["preprocessing"])
+        or ("word_sequence_length_limit" not in config["preprocessing"][TEXT])
+        or (min_99ptile_len < float(config["preprocessing"][TEXT]["word_sequence_length_limit"]))
```
[the remark this answers is not in the document]
Comment (amholler) [reply ludwig/automl/auto_tune_config.py:145]: yes

File: ludwig/utils/strings_utils.py [hunk 5]
```diff
@@ -282,7 +284,7 @@ def create_vocabulary(
     if padding_symbol in str2idx.keys():
         pad_idx = str2idx[padding_symbol]
 
-    return vocab, str2idx, str2freq, max_line_length, pad_idx, padding_symbol, unknown_symbol
+    return vocab, str2idx, str2freq, line_length_max, line_length_99ptile, pad_idx, padding_symbol, unknown_symbol
```
[the remark this answers is not in the document]
Comment (amholler) [reply ludwig/utils/strings_utils.py:287]: good point, thanks

File: ludwig/automl/auto_tune_config.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (amholler) [reply ludwig/automl/auto_tune_config.py]: added 4 unit tests

**Review by justinxzhao — APPROVED** [review]

**Pull request merged**